common/models/acl.js

```diff
@@ -456,15 +489,34 @@ module.exports = function(ACL) {
         if (err) return callback(err, null);
 
         var resolved = self.resolvePermission(effectiveACLs, req);
+        var permission = resolved && resolved.permission;
         if (resolved && resolved.permission === ACL.DEFAULT) {
-          resolved.permission = (model && model.settings.defaultPermission) || ACL.ALLOW;
+          permission = defaultPermission || ACL.ALLOW;
         }
         debug('---Resolved---');
         resolved.debug();
+
+        // set authorizedRoles in remotingContext options argument if
+        // checkAccessForContext resolved permission status is ALLOW,
+        // else set it to empty object
+        authorizedRoles = permission === ACL.ALLOW ? authorizedRoles : {};
+        setAuthorizedRoles(authorizedRoles);
         return callback(null, resolved);
       });
     });
     return callback.promise;
+
+    // helpers
+    function setAuthorizedRoles(authorizedRoles) {
```

Comment:
I am proposing to move this function out of `ACL.checkAccessForContext` and add `remotingContext` as a new parameter, so that the function does not require a closure from `ACL.checkAccessForContext` to access that reference.

There are two reasons for my proposal:
 - I personally consider such code easier to reason about, because it's easier to see what data is read/touched by `setAuthorizedRoles`
 - It's a also a (micro) performance optimisation. In the current implementation, V8 has to create a new function every time `ACL.checkAccessForContext` is called. I don't know what's the current status, but V8 was not able to optimise such short-living functions in the past.  By moving the function out, there is only one function that's called many times, and therefore V8 can optimise the heck out of it.

I would also like to propose  to use a different name, e.g. `saveAuthorizedRolesToRemotingContext`, and to  omit the `//helpers` comment.